scm/driver/bitbucketserver/content.go

```diff
@@ -0,0 +1,87 @@
+package bitbucketserver
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/drone/go-scm/scm"
+	"net/url"
+	"strings"
+)
+
+type contentService struct {
+	client *wrapper
+}
+
+func (s *contentService) Find(ctx context.Context, repo, path, ref string) (*scm.Content, *scm.Response, error) {
+	projectRepo := strings.Split(repo, "/")
+	endpoint := fmt.Sprintf("/rest/api/1.0/projects/%s/repos/%s/browse/%s?at=%s", projectRepo[0], projectRepo[1], path, ref)
+	out := new(content)
+	res, err := s.client.do(ctx, "GET", endpoint, nil, out)
+	data := stringBuffer(out.Lines)
+	return &scm.Content{
+		Path: path,
+		Data: []byte(data),
+	}, res, err
+}
+
+func (s *contentService) Create(ctx context.Context, repo, path string, params *scm.ContentParams) (*scm.Response, error) {
+	projectRepo := strings.Split(repo, "/")
+	endpoint := fmt.Sprintf("/rest/api/1.0/projects/%s/repos/%s/browse/%s", projectRepo[0], projectRepo[1], path)
+	headerParams := make(map[string]string)
+	formParams := url.Values{}
+
+	formParams.Add("message", params.Message)
+	formParams.Add("content", string(params.Data))
+	formParams.Add("branch", params.Branch)
+
+	res, err := s.client.doForm(ctx, "PUT", endpoint, headerParams, formParams, nil)
+	return res, err
+}
+
+func (s *contentService) Update(ctx context.Context, repo, path string, params *scm.ContentParams) (*scm.Response, error) {
+	projectRepo := strings.Split(repo, "/")
+	endpoint := fmt.Sprintf("/rest/api/1.0/projects/%s/repos/%s/browse/%s", projectRepo[0], projectRepo[1], path)
+	headerParams := make(map[string]string)
+	formParams := url.Values{}
+
+	formParams.Add("message", params.Message)
+	formParams.Add("content", string(params.Data))
+	formParams.Add("branch", params.Branch)
+	formParams.Add("sourceCommitId", params.Branch)
+
+	res, err := s.client.doForm(ctx, "PUT", endpoint, headerParams, formParams, nil)
+	return res, err
+}
+
+func (s *contentService) Delete(ctx context.Context, repo, path, ref string) (*scm.Response, error) {
+	return nil, scm.ErrNotSupported
+}
+
+func (s *contentService) List(ctx context.Context, repo, path, ref string, opts scm.ListOptions) ([]*scm.ContentInfo, *scm.Response, error) {
+	return nil, nil, scm.ErrNotSupported
+}
+
+func stringBuffer(lines []contentLines) string {
```

Comment:
use string join? https://golang.org/pkg/strings/#Join

Reply:
If the amount of content is large, the performance will be better using this method

Reply:
https://github.com/golang/go/blob/master/src/strings/strings.go#L416
Look at the implementation of the official go package, there will be no performance gap